frodo.js

```diff
@@ -3,46 +3,49 @@
  *
```

Comment:
SCRIPT-This function seems to have grown quite complex. Can we break it down further?